### ISSUE TITLE
RTDETR training on custom dataset giving "NAN" loss /error value after 50 epochs using AdamW optimizer

### ISSUE THREAD

**Issue by rajpalaakash** [ultralytics/ultralytics#7594]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi All,

I have been training the RTDETR-L model on our custom dataset since the past month with bbox of the objects specifically car damages as our output. I have ran around 100 experiments and have been getting better results as compared to the standard YoloV8-L or M. During all the trainings, what I have learned is RTDETR performs best with AdamW set as optimizer, whereas YOLO models perform better with SGD as optimizer. However, while training RTDETR I have to be very careful with AdamW as the losses goes to "NAN" value after a certain epochs/iteration number depending on the dataset size, nbs and no of epochs. I have seen the code and there is a condition which selects AdamW as optimizer if the no.of iterations are less that 10000 and also went through the formula. The Total no of train images in my dataset are **12016** with additional 847 as validation images.  
**My question is how can I train RTDETR with AdamW but for epochs greater than 50 without loss value going to NAN. Using the formula given for no-of-iterations in the code, I calculated 50 epochs as being the perfect no for less than 10000 iterations so as to chooes AdamW as optimizer with the below hyperparameters.** Also attaching are the training graphs for experiment 1 with exactly 50 epochs and experiment 2 with 100 Epochs. As can be seen after around 50 epochs the loss value goes to NAN for experiment 2 and the training stops all together. 

### The important parameters are highlighed here:-
_Batch Size:- 16
imgsz:- 1280
NBS:- 64
Freeze:- 10
lr0:- 0.001
lrf:- 0.01_
**EXPERIMENT 1 TRAINING GRAPHS (50 EPOCHS) :-** 
![image](https://github.com/ultralytics/ultralytics/assets/124160423/4c5265a5-a67a-4835-8a64-c242753d29e8)
![image](https://github.com/ultralytics/ultralytics/assets/124160423/9e5a8d55-ac73-4014-b274-7ce7d0c2157c)
![image](https://github.com/ultralytics/ultralytics/assets/124160423/4b476c5b-d713-4cca-87bd-c8ce2c538a09)

**EXPERIMENT 2 TRAINING GRAPHS ( 100 EPOCHS) **
![image](https://github.com/ultralytics/ultralytics/assets/124160423/1b96d3a6-ec4d-468a-93df-e551f69c4290)
![image](https://github.com/ultralytics/ultralytics/assets/124160423/4744ed8d-b7a6-4e64-acc3-f64e21de5f9f)
![image](https://github.com/ultralytics/ultralytics/assets/124160423/d27d438c-e693-41cc-bae8-5d101cc61729)

**Comment by glenn-jocher** [comment]
@rajpalaakash hello there!

It's great to hear that you've been getting promising results with RTDETR on your custom dataset. The issue of encountering "NAN" loss values after a certain number of epochs is indeed a challenge that can arise with the AdamW optimizer, especially when training for extended periods.

To address this, you might want to consider the following strategies:

1. **Learning Rate Adjustment**: Gradually reduce the learning rate as you approach the epoch where "NAN" values typically occur. This can help stabilize the training process.

2. **Gradient Clipping**: Implement gradient clipping to prevent the gradients from exploding, which is a common cause of "NAN" losses.

3. **Weight Decay Tuning**: Adjust the weight decay parameter. Sometimes, a lower weight decay can help prevent "NAN" losses when using AdamW.

4. **Optimizer Warmup**: Use a learning rate scheduler that includes a warmup phase, which can help the optimizer stabilize before reaching higher learning rates.

5. **Batch Size**: Experiment with different batch sizes. A smaller batch size can sometimes lead to more stable training, although it may affect the learning dynamics.

6. **Regularization Techniques**: Employ regularization techniques such as dropout or data augmentation to improve generalization and potentially stabilize training.

Remember to monitor your training closely and consider early stopping if you notice the loss starting to diverge. For more detailed guidance on these strategies, please refer to our documentation at https://docs.ultralytics.com.

Keep up the good work, and I hope these suggestions help you achieve stable training beyond 50 epochs with AdamW! 😊🚀

**Comment by simulacraliasing** [comment]
I have run into the same NAN problem after several iterations. In my case, with amp enabled, it is the float16 q,k multiplication in pytorch multi_head_attention_forward caused the multiplication result ("attn_output_weights") contains rows with all -inf values. After a softmax, these -inf rows became nan rows and infected loss to be nan as well after some steps.
Disable amp should solve this problem but will increase gpu memory consumption. My quick and dirty temporary solution is disbale torch.autocast when calculate this q,k multiplication by adding `with torch.autocast(device_type="cuda", enabled=False):` before https://github.com/pytorch/pytorch/blob/fff9d98e58cda4ad67fef6d0739f684ca7d4bbec/torch/nn/functional.py#L5466 and  https://github.com/pytorch/pytorch/blob/fff9d98e58cda4ad67fef6d0739f684ca7d4bbec/torch/nn/functional.py#L5469. And modify `attn_output_weights = torch.baddbmm(attn_mask, q_scaled, k.transpose(-2, -1))` to `attn_output_weights = torch.baddbmm(attn_mask, q_scaled.float(), k.transpose(-2, -1).float())`

**Comment by Nikhil-S-Shinde** [comment]
@glenn-jocher Hi can you please clarify what modifications are needed and their locations in the repo to get rid of the NaN problem when using AdamW optimizer. 
I am training Yolo with a custom backbone and RTDETR head. The architecture trains without errors on SGD optimizer.

```
train parameters:
device = 0,1
batch = 8
nbs = 64
amp = True
workers = 4
optimizer = AdamW
lr0 = 0.001
lrf=0.01
momentum=0.9
warmup_epochs = 5
cos_lr = True
```

**Comment by glenn-jocher** [comment]
For AdamW optimizer stability with RTDETR heads and AMP enabled, we recommend these steps:

1. **Disable AMP** by setting `amp=False` in your training arguments. This resolves most NaN issues as noted in the [RTDETR training documentation](https://docs.ultralytics.com/reference/models/rtdetr/train/) regarding AMP instability.

2. Verify AMP functionality using our built-in check:
```python
from ultralytics.utils.checks import check_amp
check_amp(model)
```

3. If keeping AMP enabled, implement gradient clipping via `args.grad_clip=1.0` and reduce `lr0` (try 0.0001-0.0005). 

The [configuration docs](https://docs.ultralytics.com/usage/cfg/) detail these parameters. For enterprise support, please visit our [Licensing page](https://ultralytics.com/license).
